File: src/mnemonic/parser/exe.py
"""EXEファイルからの埋め込みXP3抽出機能

Windows EXE形式のゲームファイルには、XP3アーカイブが埋め込まれていることがある。
このモジュールは、EXEファイル内のXP3オフセットを検出し、抽出する機能を提供する。
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

# XP3ファイルのマジックバイト（11バイト）
XP3_MAGIC = b"XP3\x0d\x0a\x20\x0a\x1a\x8b\x67\x01"


@dataclass(frozen=True)
class EmbeddedXP3Info:
    """EXE内埋め込みXP3の情報

    Attributes:
        offset: EXE内でのXP3開始オフセット
        estimated_size: 推定サイズ（EXE終端までのバイト数）
    """

    offset: int
    estimated_size: int


class EmbeddedXP3Extractor:
    """EXEファイルから埋め込みXP3を抽出するクラス

    使用例:
        >>> extractor = EmbeddedXP3Extractor(Path("game.exe"))
        >>> xp3_list = extractor.find_embedded_xp3()
        >>> extracted = extractor.extract_all(Path("/tmp/extracted"))
    """

    def __init__(self, exe_path: Path) -> None:
        """初期化

        Args:
            exe_path: 対象EXEファイルのパス

        Raises:
            FileNotFoundError: EXEファイルが存在しない場合
        """
        if not exe_path.exists():
            raise FileNotFoundError(f"EXEファイルが見つかりません: {exe_path}")
        self._exe_path = exe_path

    @property
    def exe_path(self) -> Path:
        """対象EXEファイルのパス"""
        return self._exe_path

    def find_embedded_xp3(self) -> list[EmbeddedXP3Info]:
        """EXE内の埋め込みXP3を検索する

        EXEファイルをスキャンし、XP3マジックバイトを検出する。
        複数のXP3が埋め込まれている場合、すべてを返す。

        Returns:
            検出されたXP3情報のリスト（オフセット順）
        """
        with open(self._exe_path, "rb") as f:
            content = f.read()

        file_size = len(content)
        offsets: list[EmbeddedXP3Info] = []

        # XP3マジックバイトをスキャンしてオフセットを記録
        pos = 0
        while True:
            offset = content.find(XP3_MAGIC, pos)
            if offset == -1:
                break
            offsets.append(EmbeddedXP3Info(offset=offset, estimated_size=0))
            pos = offset + 1

        # 推定サイズを計算（次のXP3までまたはファイル終端まで）
        result: list[EmbeddedXP3Info] = []
        for i, info in enumerate(offsets):
            if i + 1 < len(offsets):
                estimated_size = offsets[i + 1].offset - info.offset
            else:
                estimated_size = file_size - info.offset
            result.append(EmbeddedXP3Info(offset=info.offset, estimated_size=estimated_size))

        return result

    def extract_all(self, output_dir: Path) -> list[Path]:
        """検出したすべてのXP3を抽出する

        Args:
            output_dir: 抽出先ディレクトリ

        Returns:
            抽出されたXP3ファイルパスのリスト

        Raises:
            FileNotFoundError: EXEファイルが存在しない場合
        """
        output_dir.mkdir(parents=True, exist_ok=True)

        xp3_list = self.find_embedded_xp3()
        if not xp3_list:
            return []

        with open(self._exe_path, "rb") as f:
            content = f.read()

        extracted_files: list[Path] = []
        base_name = self._exe_path.stem

        for i, info in enumerate(xp3_list):
            xp3_data = content[info.offset : info.offset + info.estimated_size]
            output_file = output_dir / f"{base_name}_{i}.xp3"
            output_file.write_bytes(xp3_data)
            extracted_files.append(output_file)

        return extracted_files

File: internal/parser/exe.go
// Package parser はゲーム入力ファイル（EXE / XP3アーカイブ）の解析、
// ゲームエンジン構成の検出、アセットファイルの分類を行う。
package parser

import (
	"bytes"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// XP3Magic はXP3アーカイブの完全なマジックバイト列（11バイト）。
var XP3Magic = []byte{'X', 'P', '3', 0x0d, 0x0a, 0x20, 0x0a, 0x1a, 0x8b, 0x67, 0x01}

// ErrEXENotFound はEXEファイルが存在しない場合のセンチネルエラー。
var ErrEXENotFound = errors.New("EXEファイルが見つかりません")

// EmbeddedXP3Info はEXE内埋め込みXP3の情報を表す。
type EmbeddedXP3Info struct {
	// Offset はEXE内でのXP3開始オフセット。
	Offset int64
	// EstimatedSize は推定サイズ（次のXP3の開始位置またはEXE終端までのバイト数）。
	EstimatedSize int64
}

// EmbeddedXP3Extractor はEXEファイルから埋め込みXP3を抽出する。
//
// Windows EXE形式のゲームファイルには、XP3アーカイブが埋め込まれていることがある。
// このクラスは、EXEファイル内のXP3オフセットを検出し、抽出する機能を提供する。
type EmbeddedXP3Extractor struct {
	exePath string
}

// NewEmbeddedXP3Extractor はexePathを対象に初期化する。
//
// exePathが存在しない場合はErrEXENotFoundを返す。
func NewEmbeddedXP3Extractor(exePath string) (*EmbeddedXP3Extractor, error) {
	if _, err := os.Stat(exePath); err != nil {
		return nil, fmt.Errorf("%w: %s", ErrEXENotFound, exePath)
	}

	return &EmbeddedXP3Extractor{exePath: exePath}, nil
}

// ExePath は対象EXEファイルのパスを返す。
func (e *EmbeddedXP3Extractor) ExePath() string {
	return e.exePath
}

// FindEmbeddedXP3 はEXE内の埋め込みXP3を検索する。
//
// EXEファイルをスキャンし、XP3マジックバイトを検出する。
// 複数のXP3が埋め込まれている場合、すべてをオフセット順に返す。
func (e *EmbeddedXP3Extractor) FindEmbeddedXP3() ([]EmbeddedXP3Info, error) {
	content, err := os.ReadFile(e.exePath) //nolint:gosec // コンストラクタでexists検証済みのユーザー指定EXEパスを読む用途のため妥当
	if err != nil {
		return nil, fmt.Errorf("EXEファイルの読み込みに失敗しました: %w", err)
	}

	fileSize := int64(len(content))

	// XP3マジックバイトをスキャンしてオフセットを記録する。
	// posをoffset+1から再開することで、Python版と同様に1バイトずつ
	// 前進しながら重複を許容した探索を行う（bytes.Indexは重複開始位置を
	// 見逃さないため、offset+len(magic)からの再開よりも安全）。
	var offsets []int64
	pos := 0
	for {
		idx := bytes.Index(content[pos:], XP3Magic)
		if idx == -1 {
			break
		}
		offset := pos + idx
		offsets = append(offsets, int64(offset))
		pos = offset + 1
	}

	result := make([]EmbeddedXP3Info, 0, len(offsets))
	for i, offset := range offsets {
		var estimatedSize int64
		if i+1 < len(offsets) {
			estimatedSize = offsets[i+1] - offset
		} else {
			estimatedSize = fileSize - offset
		}
		result = append(result, EmbeddedXP3Info{Offset: offset, EstimatedSize: estimatedSize})
	}

	return result, nil
}

// ExtractAll は検出したすべてのXP3をoutputDirへ抽出する。
//
// 抽出されたファイルはexeファイル名のstemに連番を付与した名前
// （例: game_0.xp3）で保存される。
func (e *EmbeddedXP3Extractor) ExtractAll(outputDir string) ([]string, error) {
	if err := os.MkdirAll(outputDir, 0o750); err != nil {
		return nil, fmt.Errorf("出力ディレクトリの作成に失敗しました: %w", err)
	}

	xp3List, err := e.FindEmbeddedXP3()
	if err != nil {
		return nil, err
	}
	if len(xp3List) == 0 {
		return []string{}, nil
	}

	content, err := os.ReadFile(e.exePath) //nolint:gosec // コンストラクタでexists検証済みのユーザー指定EXEパスを読む用途のため妥当
	if err != nil {
		return nil, fmt.Errorf("EXEファイルの読み込みに失敗しました: %w", err)
	}

	baseName := strings.TrimSuffix(filepath.Base(e.exePath), filepath.Ext(e.exePath))

	extracted := make([]string, 0, len(xp3List))
	for i, info := range xp3List {
		data := content[info.Offset : info.Offset+info.EstimatedSize]
		outputFile := filepath.Join(outputDir, fmt.Sprintf("%s_%d.xp3", baseName, i))
		if err := os.WriteFile(outputFile, data, 0o600); err != nil { //nolint:gosec // outputFileはoutputDirとEXEファイル名由来の固定書式で構築され外部入力を含まない
			return nil, fmt.Errorf("XP3ファイルの書き込みに失敗しました: %w", err)
		}
		extracted = append(extracted, outputFile)
	}

	return extracted, nil
}
